neo4j/bookmarks.go

```diff
@@ -10,13 +35,129 @@ type BookmarkManager interface {
 	// UpdateBookmarks updates the bookmark for the specified database
 	// previousBookmarks are the initial bookmarks of the bookmark holder (like a Session)
 	// newBookmarks are the bookmarks that are received after completion of the bookmark holder operation (like the end of a Session)
-	UpdateBookmarks(database string, previousBookmarks, newBookmarks Bookmarks)
+	UpdateBookmarks(ctx context.Context, database string, previousBookmarks, newBookmarks Bookmarks)
+
+	// GetAllBookmarks returns all the bookmarks associated with the specified databases
+	GetAllBookmarks(ctx context.Context, databases []string) Bookmarks
 
 	// GetBookmarks returns all the bookmarks associated with the specified database
-	GetBookmarks(database string) Bookmarks
+	GetBookmarks(ctx context.Context, database string) Bookmarks
+}
 
-	// GetAllBookmarks returns all the bookmarks associated with the specified databases
-	GetAllBookmarks(databases []string) Bookmarks
+type BookmarkManagerConfig struct {
+	// Initial bookmarks per database
+	InitialBookmarks map[string]Bookmarks
+
+	// Hook providing bookmarks for a given database
+	// The results returned by this hook are always included in BookmarkManager#GetBookmarks
+	// This is useful for users storing bookmarks externally e.g.
+	BookmarkSupplier func(string) Bookmarks
+
+	// Hook called whenever bookmarks for a given database get updated
+	// The hook is called with the database and the new bookmarks
+	BookmarkUpdateNotifier func(string, Bookmarks)
+}
+
+type bookmarkManager struct {
+	bookmarks       map[string]collection.Set[string]
+	locks           map[string]racing.ReadWriteMutex
+	supplierFn      func(string) Bookmarks
+	notifyUpdatesFn func(string, Bookmarks)
+	globalLock      racing.Mutex
+}
+
+func (b *bookmarkManager) UpdateBookmarks(ctx context.Context, database string, previousBookmarks, newBookmarks Bookmarks) {
+	databaseLock := b.findOrInitDatabaseLock(ctx, database)
+	if !databaseLock.TryWriteLock(ctx) {
+		//TODO: better error handling?
+		panic("could not acquire write lock for updated database bookmarks")
+	}
+	defer databaseLock.UnlockWrite()
+
+	currentBookmarks, found := b.bookmarks[database]
+	if !found {
+		b.bookmarks[database] = collection.NewSet(newBookmarks)
+		return
+	}
```

Comment:
You should notify the current bookmarks also here.

Reply:
obrigado

Reply:
Fixed in e628a710dceaa92357a96caac64e4768efbc30da